### ISSUE TITLE
[Wiesbaden] Data

### ISSUE THREAD

**Issue by jklmnn** [offenesdresden/ParkAPI#176]
Wiesbaden also has parking data available at https://www.wiesbaden.de/leben-in-wiesbaden/verkehr/verkehrsinfos/parken.php.
The raw data seems to be at https://wi.memo-rheinmain.de/wiesbaden/parkliste.phtml?order=carparks. While it looks quite ugly to parse it seems to have all the data we need.

**Comment by kiliankoe** [comment]
> While it looks quite ugly to parse

Lol, this is *far* better than most others 😄 

<img width="434" alt="Screen Shot 2019-04-04 at 10 12 54 1" src="https://user-images.githubusercontent.com/2625584/55539995-469f4d00-56c2-11e9-8bfb-da530a64c601.png">

**Comment by jklmnn** [comment]
But only with whitespace correction :smile: 
![image](https://user-images.githubusercontent.com/7999005/55540113-8cf4ac00-56c2-11e9-863a-d31419c15cc2.png)

**Comment by AugustQu** [comment]
Hi,

here I am.
From what I understand: first I deliver the coordinates of the parking-lots given in the URL above.

So this may look like:
`static_data["City I"] = {'latitude': "50.08121",
                          'longitude': "8.23637",
                          'adresse': "Schwalbacher Str. 41-43"}
`

Am I correct?

And a question:
this parking-lot is closed on sundays/holidays and opened on monday to saturday from 7:00 to 20:00. How do I enter this information?

PS: I've taken this file as an example:
https://github.com/Bytespeicher/pyParkingLotEF2Json/blob/master/pyParkingLotEF2Json.py

**Comment by AugustQu** [comment]
I've seen a lot of py-files. Looking at the code I found:

- create a loop over all parking-lots (14 for WI)
- search for the name (td-Element)
- next is the 'Freie Stellplätze'
- finally: return the name together wit the numbers.

Am I right?

PS: I'm new to python.

**Comment by jklmnn** [comment]
> Hi,
> 
> here I am.
> From what I understand: first I deliver the coordinates of the parking-lots given in the URL above.
> 
> So this may look like:
> `static_data["City I"] = {'latitude': "50.08121", 'longitude': "8.23637", 'adresse': "Schwalbacher Str. 41-43"} `
> 
> Am I correct?
> 
> And a question:
> this parking-lot is closed on sundays/holidays and opened on monday to saturday from 7:00 to 20:00. How do I enter this information?
> 
> PS: I've taken this file as an example:
> https://github.com/Bytespeicher/pyParkingLotEF2Json/blob/master/pyParkingLotEF2Json.py

The project you are referring to seems to be outdated. Also it doesn't use our API.

The files relevant for you reside in [park_api/cities](https://github.com/offenesdresden/ParkAPI/tree/master/park_api/cities). I'd recommend you to take a look at [`Sample_City.py`](https://github.com/offenesdresden/ParkAPI/blob/master/park_api/cities/Sample_City.py) and [`Sample_City.geojson`](https://github.com/offenesdresden/ParkAPI/blob/master/park_api/cities/Sample_City.geojson). You could copy those files to `Wiesbaden.py` and `Wiesbaden.geojson` respectively and then edit them. Wiesbaden should then automatically appear in your local instance.

For parsing we use beautifulsoup4. I'd advise you to do the same since it is easily powerful enough for the task and already a dependency. About the data the relevant table is accessible via the xpath `/html/body/table/tbody/tr[3]/td/div/table/tbody` and then iterate over all `tr`s except the first one (which is the table header) and get the data from the `td`s.

Another far simpler, but important and time consuming task would be filling the geojson file which needs to be done by hand but only once. The most important information are the coordinates.

**Comment by AugustQu** [comment]
Hi,

I'm not only new to Python but also new to Github.

I copied the whole project as a ZIP-file, expanded it and modified the file _Sample_city.geojson_ adding all the 14 parking-lots of Wiesbaden. Finally I saved it as _Wiesbaden.geojson_, locally on my PC.

Some questions:
#1: in the 'features' of this file coordinates have to be entered. I entered the coordinates of the city town hall. Is this correct?
#2: in the 'properties' I have to enter an URL. Here I entered "https://www.wiesbaden.de/leben-in-wiesbaden/verkehr/verkehrsinfos/parken.php". Correct?
Also as a source I entered: "https://wi.memo-rheinmain.de/wiesbaden/parkliste.phtml?order=carparks". Correct?
#3: there is one parking lot that has 2 entries for cars. Can this be handled? Or shall I omit the second entry?

**Comment by jklmnn** [comment]
I'd recommend to check out the [GitHub Hello World](https://guides.github.com/activities/hello-world/) tutorial and the [git tutorial](https://git-scm.com/docs/gittutorial) (while GitHub builds upon git, git is a free version control system). There also is a [collection](https://try.github.io/) of resources that looks like a good start.

As for the general workflow, you should fork this repository, do your changes on your on fork and the create a pull request with your changes for us to review.

1. The data includes a list of features with the first being the city itself and all others being parking lots. The coordinates of the first feature should be either the geographical center of the city or the geographical center of the distribution of parking lots. We usually take the coordinates of large map providers for this.
2. Yes that's both correct. The `url` field denotes the public accessible location of the data (city homepage, etc) and the `source` denotes the actual source of the information that the parser will receive.
3. Do you mean entries in how you enter the lot or data entries? In the first case we only use one address and coordinate. In the second case I'm not sure how it would look like.

**Comment by AugustQu** [comment]
playing with Python I got this result:
my code:

> table=soup.select('table')
td = table[2].find_all("td")
i=0
while i < len(td):
    print('name:',td[i+1].text.strip())
    print('frei=', int(td[i+2].text.split()[0]))
    print('total=', int(td[i+2].text.split()[2]))
    i += 5

the result, first entry only:

> ('name:', u'City I')
('frei=', 137)
('total=', 165)


hint: I added the html statically for my test.

So I have to add these informations to the data-object and I should be done.

Looking into the Sample_City.py I find these lines:

>         data["lots"].append({
            "name": lot.name,
            "free": lot_free,
            "total": lot_total,
            "address": lot.address,
            "coords": lot.coords,
            "state": state,
            "lot_type": lot.type,
            "id": lot.id,
            "forecast": False,

The first 3 lines are the information given above, the rest comes from the geojson-file looked into via the name.
Correct?

PS: I have to check for state and lot.id.

**Comment by AugustQu** [comment]
I just added the files _Wiesbaden.geojson_ and _Wiesbaden.py_ but I'm not sure that I did it correctly.

In the Python-script I added 4 TODOs which describe lines to look for:

- the handling of last_updated
- the state (twice)
- the lot.id which I do not know

Can you please check these lines and also the whole script.
Thanks.

**Comment by AugustQu** [comment]
one error I made:
I loaded the file Sample_city.geojson, replaced the contents with my file (Wiesbaden.geojson) and saved it as Wiesbaden.geojson.
Instead I should have created a new file via the button.

Same for the py-file.

Sorry.

**Comment by jklmnn** [comment]
Sorry for my late reply.

> The first 3 lines are the information given above, the rest comes from the geojson-file looked into via the name.
Correct?

Yes that's true.

There are some issues about the pull requests. One is they're opened against master which we won't merge until everything works perfectly. But you can easily choose the `wiesbaden` branch with the dropdown on the left:
![image](https://user-images.githubusercontent.com/7999005/55942269-57ad0880-5c44-11e9-8b2c-5bf8a2a2cbff.png)

The second thing is I'd suggest you only make a single pull request with all your changes. This helps us to directly comment on all parts of your code to help you better. By pushing to the branch of the single pull request it will be updated and we always will see the changes.

**Comment by AugustQu** [comment]
what do I have to do now? Add a new Pull-request against the Wiesbaden-branch?

Or do you handle this situation?

**Comment by jklmnn** [comment]
Yes, you have to open a new pull request against the wiesbaden branch.

**Comment by AugustQu** [comment]
OK. I added the files _Wiesbaden.py_ and _Wiesbaden.geojson_ in the cities-directory of the Wiesbaden-branch and did a Pull-request. I hope this is correct now.

Unfortunately I added the file _Wiesbaden.py_ to the wrong directory in the first try so please delete this file. And also it looks like I did overwrite the file _Sample_city.py_ with the code for Wiesbaden. Can I correct this? Or do you simply reject the 'modifications'?

**Comment by jklmnn** [comment]
Thanks, that looks better. Since the overwritten `Sample_City.py` file is in its own commit it doesn't pose a problem as I can omit it when merging.

**Comment by AugustQu** [comment]
I will look at the status of the parking-lot over the weekend, then make some changes to the code.

I've seen that the data on wiesbaden.de showed the status of 2 parking-lots as closed, I saw this last night. I will try to find out how they do it.

**Comment by AugustQu** [comment]
it looks like there are only minor modifications needed. So I will look for the status, make the needed changes plus the already mentioned modifications and deliver them together.

I gues the next version will come on monday.

One hint: the word _Uhr_ is included in the following line taken from the file Sample_city.py:

>         "last_updated": convert_date(last_updated, "%d.%m.%Y %H:%M Uhr"),

**Comment by jklmnn** [comment]
> One hint: the word Uhr is included in the following line taken from the file Sample_city.py

This is of course just an example and needs to be adapted to the actual data.

**Comment by AugustQu** [comment]
OK, I got it.

Uploading my code to the branch wiesbaden I did encounter 2 problems:

- I do not see the old files any more

- so I tried to create new files but could not save them because I have no write access

What can I do now?

**Comment by AugustQu** [comment]
there is this: AugustQu/ParkAPI
there is a wiesbaden-branch where I can find my files.

I will add my latest modifications to these files and then make another pull-request.

**Comment by AugustQu** [comment]
so I hope I did it correctly.

**Comment by AugustQu** [comment]
I made some modifications to the py-script and will upload this file later. These modifications were of the type beautifying.

But I do see one problem: there is a parking-lot with the name Coulinstraße. On the memo-rheinmain-page it appears as: Coulinstra**ÃŸ**e, in the geojson-file I added it as Coulinstra**ss**e. In the py-script this will be handled as:
`        parking_name = td[i+1].text.strip()
        lot = geodata.lot(parking_name)
`
so the name is extracted from the memo-rheinmain-page and this is used for the search in the geojson-file. This will result in no match.

How can I handle this? Enter the name in the geojson-file as Coulinstra**ß**e?

**Comment by AugustQu** [comment]
I found a typing-error:
in the geojson-file the name of a parking-lot is written as: Rhei**n M**ain
in the memo-rheinmain-site it is written as: Rhei**nM**ain

I will fix this.

Put this aside: does the script work?

**Comment by jklmnn** [comment]
Sorry for my late reply, I'm quite busy at the moment.

> so I hope I did it correctly.
It seems you pushed your changes to the already existing pull request, so yes that's correct. I will review it when I find time.

About ß in street names, they should be usually encoded in UTF-8 that supports those characters. Although if the page doesn't have the correct encoding they will be erroneous. I think we have a solution for this to at least convert those characters to ASCII (@kiliankoe do you remember how we fixed that in other cities?).

> Put this aside: does the script work?

Same as above, I can tell when I find time to review (and test ofc.).

**Comment by AugustQu** [comment]
> Sorry for my late reply, I'm quite busy at the moment.

OK. The usual thing: I created some lines and want to see the result immediately.....

I've jut added some minor modifications to the two files. One ist the name 'Coulinstra**ß**e' in the geojson-file and the other is a bit of cleaning the code in the py-script.

**Comment by jklmnn** [comment]
Thanks for the changes, I reviewed them. There are two stray lines in the geojson that shouldn't be there. If you remove them I think it's ready to be merged.

One last thing. There is the `active_support` flag in the json which decides if the city is enabled by default in the apps. If you want the city to be enabled by default I'd suggest you set it to `true`.

If the website changes and the parser breaks (happens usually rarely) we will either try to fix it or (if the fix requires more work) we probably set active_support to false again unless someone fixes it. That what the flag is for.

**Comment by kiliankoe** [comment]
> I think we have a solution for this to at least convert those characters to ASCII (@kiliankoe do you remember how we fixed that in other cities?).

We have [`util.remove_special_chars`](https://github.com/offenesdresden/ParkAPI/blob/master/park_api/util.py#L55), but that doesn't appear to be sophisticated enough for this specific use-case. There must be a default and sane way to deal with this, probably even built-in to Python.

**Comment by AugustQu** [comment]
OK. I've created another pull-request.

I set `active_support` to `true`. And I also removed a lot of lines (245 in my copy, 440 here in Github), a copy-paste-error.

Open topic: the sz = ß. It's somehow coded on the web-site. If you can convert this to another (ASCII-)character then I have to change the geojson-file accordingly. I will do that, please let me know.

**Comment by AugustQu** [comment]
regarding the sz-problem:
can you please simply do a test. Maybe ist's already working without any problems.

I just did fetch the page via wget and looked into it. The name of the problematic parking-lot looked like:  Coulinstra**ß**e, so this would be OK.

If I look at it via my browser here I see it as Coulinstra**ÃŸ**e
So, maybe, it's a problem with the settings of my browser.

**Comment by jklmnn** [comment]
There is no need to create a new pull request each time. Pushing to the already open one is sufficient. I closed the second pull request as the original one contains all the changes we need.

I took a look at the HTML. The data itself seems to be encoded in UTF-8 correctly but the browser doesn't seem to know. Adding a `<meta charset="utf-8"/>` manually at the top of the HTML fixes the issue in the browser. Since we want to have things encoded in UTF-8 there's no need to change anything here.
@kiliankoe we should probably check how encoding is handled in general in our JSON and if and how we could add encoding information.

From your side there's nothing to do about this issue.

**Comment by AugustQu** [comment]
OK. I will now wait for the next step.

**Comment by AugustQu** [comment]
Thanks!

Can't wait to see it 'live'.

PS: this script was my first work in Python......

**Comment by jklmnn** [comment]
Wiesbaden is live now: https://parkendd.de/map.html#Wiesbaden
Please take a short look if everything is correct, especially coordinates.

If you have a Twitter account I would mention you when I tweet about this.

**Comment by AugustQu** [comment]
Hi,

THANKS! Looks good.

A short test showed up one problem: Coulinstra**ß**e. The sz-problem. And with the name fetched from the web-page there is no match in the geojson-file, so no coordinates are availabel.

**#1**: what can we do to solve this? Is there a general solution?

**#2**: I can do a quick-fix, an ugly one. Something linke this:
`if 'Coulinstr' in parking_name then parking_name='Couloinstraße'`
or so.

Please let me know.

Regarding publication: I will write a text in my blog presenting this app. And maybe another text about the 'making of' of this app. For this I need your help.

**Comment by AugustQu** [comment]
OK, changed the code (and added some comments). It seems to work here on my site.

What can I do now? Modify the file Wiesbaden.py in my branch (AugustQu/ParkApi) and start a new pull-request?

**Comment by jklmnn** [comment]
I'll reopen this issue until this problem is fixed.

Thanks for the fix. Please [update your fork](https://help.github.com/en/articles/syncing-a-fork) since we've also pushed some changes and then create a new pull request with the fix.

**Comment by AugustQu** [comment]
Done.

**Comment by jklmnn** [comment]
I've seen you updated your master branch but not your wiesbaden branch. So if you want to do a clean pull request from your wiesbaden branch against our wiebaden branch you have to do the same with the wiesbaden branches as you did with master. But I forgot to tell this in my previous comment, so no worries.

This time I can manually merge the changes as they're small so you don't need to do anything.

**Comment by jklmnn** [comment]
And thanks for your help!

**Comment by AugustQu** [comment]
> I've seen you updated your master branch but not your wiesbaden branch.

I looked into my branch and do not see a file Wiesbaden.py in my master branch, but I do see it in my wiesbaden-branch.

And the pull-request is made from AugustQu/wiesbaden against offenesDresden/wiesbaden.

Can you please check this again.

**Comment by AugustQu** [comment]
regarding whitespace:

my editor here does not show this. And is your system so sensible?

**Comment by AugustQu** [comment]
whitespace:
in my file here there is no whitespace at the end of that line. So (again) it looks like a copy-paste-problem.

**Comment by jklmnn** [comment]
Trailing whitespace doesn't necessarily break anything on the running server. But it is at least unclean code which we want to prevent as much as possible and it may break data that spans over multiple lines if ignored.
Related: https://softwareengineering.stackexchange.com/a/121560

Also it generates changes in files that are no longer easily visible creating different versions of the same software that have no visible change. Example:
![image](https://user-images.githubusercontent.com/7999005/56792807-beb4ea80-680a-11e9-910d-4c6a73602fa9.png)
In this case a trailing whitespace has been introduced but for the reader it is not clear what has changed.
Also there are editor(-plugins) that show trailing whitespace. When I open this file in Vim the whitespace is marked in red:
![image](https://user-images.githubusercontent.com/7999005/56793074-67fbe080-680b-11e9-9773-106ead664842.png)

**Comment by AugustQu** [comment]
> But it is at least unclean code 

OK.

I have used gedit for editing the script but I can switch to eclipse which shows whitespace.

Nevertheless this whitespace was entered by copying the lines from my file to the file here.....

**Comment by AugustQu** [comment]
3 simple questions:

#1: in the app I've seen lines in green, yellow, red and blue.
green lines mean a lot of free space - OK.
yellow lines mean percentage has fallen under a value of ?? (which value is it?)
red lines mean only some parking spots are available (from which percentage on the color will change?)
blue line means ?????

#2: I've seen texts like 'no data availble'. Will this text appear when my code sets `parking_state = 'nodata' `?

#3: closed parking-lots disappear from the list. Is this the intended behaviour?

**Comment by jklmnn** [comment]
I assume you talk about the Android app which I wrote. I can't tell you about the iOS and Windows apps.

1. green: 100% free - 20% free
    yellow: 20% free - 5% free
    red: <5% free, closed
    blue: no data available
2. yes
3. If you tap the three dots in the upper right corner and select settings you get to the settings dialog. There you can choose which parking lots are shown. By default full (0% free) and closed lots are hidden.

**Comment by jklmnn** [comment]
I checked the map and it seems to work now, thanks for the fix.

**Comment by AugustQu** [comment]
YES! Looks good.

I will write a text in my blog about this app but this will take some time......

**Comment by AugustQu** [comment]
> I assume you talk about the Android app 

Yes. And thanks for the explanations.

**Comment by AugustQu** [comment]
another question:

in the top-line of the app there is a world-icon. Tipping on it shows some circles in green/red.

What is this good for?

Should it show the location of the parking-lots via OpenStreetMap? On my Smartphone I only see a background in grey.

**Comment by jklmnn** [comment]
The OpenStreetMap library the app uses requires the permission to store/access local data to store its tiles. Depending on your Android version the app should have asked for that permission. If not you might enable this permission for the app in your system settings. If this permission is not given the tiles sometimes don't show up.

The apps source code is available at https://github.com/jklmnn/ParkenDD. So you can take a look at it if you want to (and know how Android apps are built). Unfortunately I don't have the time to maintain the app anymore (I only fix breaking bugs that lead to crashes/incompatibilities with newer Android versions).

**Comment by AugustQu** [comment]
here is the announcement of the availability of Wiesbaden-data on my blog:
https://linux-fuer-wi.blogspot.com/2019/05/parkendd.html

**Comment by jklmnn** [comment]
The article looks nice. I have only one issue. The city that is shown after starting the (Android) app is the closest city if location data is available or Dresden if location data isn't available. So for you if you have a location Wiesbaden should be selected by default.

**Comment by jklmnn** [comment]
I finally found some time to write a short tweet: https://twitter.com/parken_dd/status/1127983797667020801

**Comment by AugustQu** [comment]
Thanks.

**Comment by jklmnn** [comment]
It seems Wiesbaden has moved it's parking information to https://www.wiesbaden.de/leben-in-wiesbaden/verkehr/auto/parkhaeuser.php. The map consists of different parking lots, each linking to their own site with the relevant information. I don't have an idea how we would use this at the moment. I'll reopen this issue and disable Wiesbaden until we find a solution since the old source doesn't seem to be available anymore.

**Comment by AugustQu** [comment]
OK. I will look at it.

**Comment by AugustQu** [comment]
somehow I found this link: https://mobileapps.ffh.de/iphone/parkhaeuser/parkhaeuser.json?city=wiesbaden

There are 2 problems with it:
#1: can we use it?
#2: there are differences in the number: the wiesbaden-site reports 170 free parking lots for one location, the link reports 188 for the same location.

it looks like sites like:
https://www.ffh.de/verkehr/parkhaeuser/parkhaus-info-wiesbaden.html
https://www.harmonyfm.de/service/parkhaeuser/wiesbaden.html
https://www.planetradio.de/service/parkhaeuser/wiesbaden.html

use the same data-source.

**Comment by jklmnn** [comment]
The data source looks good on a quick view. After all it's JSON so much better than a website. The data difference may be caused by their update intervals (e.g. the data could be 5-10 minutes old which is reasonable for such a difference). We should check that, maybe there is a timestamp to compare somewhere.
So I think we should use that as it currently seems to be the best option.

**Comment by AugustQu** [comment]
there is a line in the JSON-file saying:

statusTime | "21.05.2021, 15:51 Uhr"

To me the data looks like very old. I will check this over the day.

And I will try to contact someone in Wiesbaden over the weekend.

**Comment by AugustQu** [comment]
ich habe dazu mal etwas geschrieben:
https://linux-fuer-wi.blogspot.com/2021/06/open-data-worte-und-taten.html

**Comment by AugustQu** [comment]
es sind etliche Monate vergangen, aber am Status hat sich nichts geändert. Hier ist meine Beschreibung samt Kommentar dazu:
https://linux-fuer-wi.blogspot.com/2021/10/parkendd-weiterhin-ohne-wiesbaden.html

Ich will dran bleiben. Aber, wie sagt man bei uns im Dorf: da kannste ach em Oxx ins Horn petze.

**Comment by jklmnn** [comment]
From what I saw at least it looks like they aim to provide a working solution in the near future (early next year).

**Comment by AugustQu** [comment]
Yes, someday they will offer a new solution. Maybe.

Btw: currently they have a lot of problems, keyword Salzbachtalbrücke.

**Comment by AugustQu** [comment]
maybe things will change (a big maybe).

Please look at this document: https://piwi.wiesbaden.de/dokument/1/2818223

Coming tuesday this event will take place: https://piwi.wiesbaden.de/sitzung/detail/2718746/tagesordnung/oeffentlich

We will see whats happening then. I will be there and listen to the discussion.

**Comment by kiliankoe** [comment]
Thanks for being so involved!
